libcudacxx/include/cuda/std/__cmath/min_max.h

```diff
@@ -21,224 +21,131 @@
 {
-#  if defined(_CCCL_BUILTIN_FMAX)
-  return _CCCL_BUILTIN_FMAXL(__x, __y);
-#  else // ^^^ _CCCL_BUILTIN_FMAX ^^^ / vvv !_CCCL_BUILTIN_FMAX vvv
-  return ::fmaxl(__x, __y);
-#  endif // !_CCCL_BUILTIN_FMAX
-}
-#endif // _CCCL_HAS_LONG_DOUBLE()
-
 #if _LIBCUDACXX_HAS_NVFP16()
-[[nodiscard]] _CCCL_API inline __half fmax(__half __x, __half __y) noexcept
-{
-  NV_IF_ELSE_TARGET(NV_IS_DEVICE,
-                    (return ::__hmax(__x, __y);),
-                    (return __float2half(_CUDA_VSTD::fmaxf(__half2float(__x), __half2float(__y)));))
-}
-template <class _A1, enable_if_t<is_arithmetic_v<_A1>, int> = 0>
-[[nodiscard]] _CCCL_API inline __promote_t<float, _A1> fmax(__half __x, _A1 __y) noexcept
-{
-  return _CUDA_VSTD::fmaxf(__half2float(__x), __y);
-}
-
-template <class _A1, enable_if_t<is_arithmetic_v<_A1>, int> = 0>
-[[nodiscard]] _CCCL_API inline __promote_t<_A1, float> fmax(_A1 __x, __half __y) noexcept
-{
-  return _CUDA_VSTD::fmaxf(__x, __half2float(__y));
-}
+  if constexpr (is_same_v<_Tp, __half>)
+  {
+    NV_IF_ELSE_TARGET(NV_IS_DEVICE,
+                      (return ::__hmax(__x, __y);),
+                      (return ::__float2half(_CUDA_VSTD::fmax(::__half2float(__x), ::__half2float(__y)));))
+  }
 #endif // _LIBCUDACXX_HAS_NVFP16()
-
 #if _LIBCUDACXX_HAS_NVBF16()
-[[nodiscard]] _CCCL_API inline __nv_bfloat16 fmax(__nv_bfloat16 __x, __nv_bfloat16 __y) noexcept
-{
-  NV_IF_ELSE_TARGET(NV_IS_DEVICE,
-                    (return ::__hmax(__x, __y);),
-                    (return __float2bfloat16(_CUDA_VSTD::fmaxf(__bfloat162float(__x), __bfloat162float(__y)));))
-}
-template <class _A1, enable_if_t<is_arithmetic_v<_A1>, int> = 0>
-[[nodiscard]] _CCCL_API inline __promote_t<float, _A1> fmax(__nv_bfloat16 __x, _A1 __y) noexcept
-{
-  return _CUDA_VSTD::fmaxf(__bfloat162float(__x), __y);
-}
-
-template <class _A1, enable_if_t<is_arithmetic_v<_A1>, int> = 0>
-[[nodiscard]] _CCCL_API inline __promote_t<_A1, float> fmax(_A1 __x, __nv_bfloat16 __y) noexcept
-{
-  return _CUDA_VSTD::fmaxf(__x, __bfloat162float(__y));
-}
+  if constexpr (is_same_v<_Tp, __nv_bfloat16>)
+  {
+    NV_IF_ELSE_TARGET(NV_IS_DEVICE,
+                      (return ::__hmax(__x, __y);),
+                      (return ::__float2bfloat16(_CUDA_VSTD::fmax(::__bfloat162float(__x), ::__bfloat162float(__y)));))
+  }
 #endif // _LIBCUDACXX_HAS_NVBF16()
-
-template <class _A1, class _A2, enable_if_t<is_arithmetic_v<_A1> && is_arithmetic_v<_A2>, int> = 0>
-[[nodiscard]] _CCCL_API inline __promote_t<_A1, _A2> fmax(_A1 __x, _A2 __y) noexcept
-{
-  using __result_type = __promote_t<_A1, _A2>;
-  static_assert(!(is_same_v<_A1, __result_type> && is_same_v<_A2, __result_type>), "");
-  return _CUDA_VSTD::fmax((__result_type) __x, (__result_type) __y);
+  if constexpr (!is_integral_v<_Tp>)
+  {
+    if (_CUDA_VSTD::isnan(__x))
+    {
+      return __y;
+    }
+    else if (_CUDA_VSTD::isnan(__y))
+    {
+      return __x;
+    }
+  }
```

Comment:
This would be the place where we could support arbitrary floating point types by casting them to `float` do we want that?

Reply:
Actually the code is guarded by `_LIBCUDACXX_HAS_NV[BF,HP]16`, so it is only available at 12.2 but we should fix that